returnn/tf/layers/basic.py

```diff
@@ -3107,29 +3118,24 @@ def get_out_data_from_opts(cls, name, axes, keep_order=NotSpecified,
       keep_order = True if BehaviorVersion.get() >= 6 else False
     assert not out_type, "currently ignored"
     input_data = get_concat_sources_data_template(sources)
+    axes = cls._get_merge_axes(
+      axes=axes, keep_order=keep_order, input_data=input_data, name="%s layer %r" % (cls, name))
     data = input_data.copy(name="%s_output" % name)
-    axes = input_data.get_axes_from_description(axes)
     if len(axes) <= 1:
       return data
-    axes = sorted(axes)
     import numpy
     res_dim = None
     if all([data.batch_shape[i] is not None for i in axes]):
       res_dim = int(numpy.prod([data.batch_shape[i] for i in axes]))
-    merge_dim_tags = [tag for (i, tag) in enumerate(data.dim_tags) if i in axes]
+    merge_dim_tags = [data.dim_tags[axis] for axis in axes]
     merge_target_axis = cls._get_target_axis(input_data=data, merge_axes=axes)
-    if any(tag.is_batch_dim() for tag in merge_dim_tags):
-      res_dim_tag_kind = Dim.Types.Batch
-    elif any(tag.is_feature_dim() for tag in merge_dim_tags):
-      res_dim_tag_kind = Dim.Types.Feature
-    else:
-      res_dim_tag_kind = Dim.Types.Spatial
```

Comment:
We might need to keep the `res_dim_tag_kind` logic.

Reply:
Either this, or we extend the dim arithmetic in general via `_OpMultTerm/_OpLinearTerm` to set the dim tag kind like here. What do you think?

Reply:
Currently `_OpMultTerm.as_dim()` uses `self.base_term().kind`, where `base_term()` returns `self.terms[-1]`. So this is kept simple by intention.

But yea, maybe it makes sense to redefine this via the logic from here.

Reply:
I implemented this in `_Op*Term.as_tag()` now.